privacyidea/lib/config.py

```diff
@@ -82,25 +84,28 @@ def __call__(cls, *args, **kwargs):
         return cls._instances[cls]
 
 
-class ConfigClass(with_metaclass(Singleton, object)):
-    """
-    The Config_Object will contain all database configuration of system
-    config, resolvers and realm.
-    It will be created at the beginning of the request and is supposed to stay
-    alive unchanged during the request.
+class SharedConfigClass(object):
     """
+    A shared config class object is shared between threads and is supposed
+    to store the current configuration with resolvers and realms, along
+    with the timestamp of the configuration.
 
+    The method ``reload_from_db()`` compares this timestamp against the
+    timestamp in the database (while taking the PI_CHECK_RELOAD_CONFIG
+    setting into account). If the database timestamp is newer, the current
+    configuration is updated.
+
+    However, app code must not access the config stored in the shared object!
+    Instead, it must use ``clone()`` or ``reload_and_clone()`` to retrieve
+    a ``LocalConfigClass`` object which holds a local configuration snapshot.
+    """
     def __init__(self):
-        """
-        Create a config object from the whole database, tables config,
-        resolver and realm.
-        """
+        self._config_lock = threading.Lock()
```

Comment:
It's true that we only do reads most of the time, but I added this lock to ensure two things:
* The ``SharedConfigClass`` object is shared between all threads. When one thread calls ``SharedConfigClass.reload_and_clone()`` to get a request-local config object, we need to make sure that no other thread updates the shared config object *while we are cloning it*: If this happens, we would end up with a config that is partly old and partly new! To avoid that, we lock the shared config object while we are cloning.
* Similarly, it could happen that two threads A and B update the ``SharedConfigClass`` simultaneously. Without a lock, it could happen that we get a config that is partly from A and partly from B. The lock ensures that we either get A's complete config, or B's complete config.